kpi/mixins/object_permission.py

```diff
@@ -320,12 +310,27 @@ def _recalculate_inherited_perms(
                 new_permission.inherited = True
                 new_permission.uid = new_permission._meta.get_field(
                     'uid').generate_uid()
-                if return_instead_of_creating:
-                    objects_to_return.append(new_permission)
-                else:
-                    new_permission.save()
+                objects_to_return.append(new_permission)
         if return_instead_of_creating:
+            if not stale_already_deleted:
+                self.permissions.filter(inherited=True).delete()
             return objects_to_return
+        else:
+            existing_objs = self.permissions.filter(inherited=True)
+            existing = existing_objs.values('permission_id', 'user_id')
+            to_add = [
+                perm
+                for perm in objects_to_return
+                if {'permission_id': perm.permission_id, 'user_id': perm.user_id}
+                not in existing
+            ]
+            for perm in objects_to_return:
+                existing_objs = existing_objs.exclude(
+                    permission_id=perm.permission_id, user_id=perm.user_id
+                )
+            existing_objs.delete()
+            for perm in to_add:
+                perm.save()
```

Comment:
What about use `bulk_create` with ignore conflicts and update field? To save more queries?